app/src/main/java/com/xmartlabs/template/controller/DbFlowController.java

```diff
@@ -0,0 +1,118 @@
+package com.xmartlabs.template.controller;
+
+import android.support.annotation.CheckResult;
+import android.support.annotation.NonNull;
+
+import com.annimon.stream.Stream;
+import com.raizlabs.android.dbflow.config.FlowManager;
+import com.raizlabs.android.dbflow.sql.language.Delete;
+import com.raizlabs.android.dbflow.sql.language.SQLCondition;
+import com.raizlabs.android.dbflow.sql.language.SQLite;
+import com.raizlabs.android.dbflow.sql.language.Select;
+import com.raizlabs.android.dbflow.sql.language.property.Property;
+import com.raizlabs.android.dbflow.structure.BaseModel;
+import com.xmartlabs.template.database.AppDataBase;
+import com.xmartlabs.template.model.DatabaseModel;
+
+import java.util.List;
+
+import io.reactivex.Completable;
+import io.reactivex.Maybe;
+import io.reactivex.Single;
+import io.reactivex.schedulers.Schedulers;
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+public class DbFlowController<Id, D extends DatabaseModel<Id>> extends Controller
+    implements DatabaseProvider<Id, D, SQLCondition> {
+  @Getter(AccessLevel.PRIVATE)
+  final Class<D> modelClass;
+  @Getter(AccessLevel.PRIVATE)
+  final Property<Id> idProperty;
+
+  @CheckResult
+  @NonNull
+  @Override
+  public Single<List<D>> getEntities(@NonNull SQLCondition... conditions) {
+    return Single
+        .fromCallable(() ->
+            SQLite.select()
+                .from(getModelClass())
+                .where(conditions)
+                .queryList()
+        )
+        .subscribeOn(Schedulers.io());
+  }
+
+  @CheckResult
+  @NonNull
+  @Override
+  public Single<List<D>> updateEntities(List<D> newEntities, @NonNull SQLCondition... conditions) {
+    return Single
+        .fromCallable(() -> {
+          FlowManager.getDatabase(AppDataBase.class)
+              .executeTransaction(databaseWrapper -> {
+                SQLite.delete(getModelClass())
+                    .where(conditions)
+                    .execute();
+                Stream.of(newEntities)
+                    .forEach(BaseModel::save);
+              });
+          return newEntities;
+        })
+        .subscribeOn(Schedulers.io());
+  }
+
+  @CheckResult
+  @NonNull
+  @Override
+  public Maybe<D> getEntity(Id id) {
+    return Maybe
+        .fromCallable(() -> new Select()
+            .from(getModelClass())
+            .where(getIdProperty().eq(id))
+            .querySingle())
+        .subscribeOn(Schedulers.io());
+  }
+
+  @CheckResult
+  @NonNull
+  @Override
+  public Single<D> createEntity(@NonNull D entity) {
+    return Completable.fromAction(entity::insert)
```

Comment:
does the insert returns a new value for the entity? I think that in some cases after you insert a database record it can change, for example if you have an autonumerated ID. I would expect that after the insert (the same for updating) the records that can be changed are updated to the returned entity.

Thoughts?